### PR TITLE
Refpix subarrays

### PR DESCRIPTION
**Pull request by stscirij**
Addresses #1652

### CODE REVIEW

File: jwst/refpix/reference_pixels.py [hunk 1]
```diff
@@ -181,6 +184,56 @@ def sigma_clip(self, data, dq, low=3.0, high=3.0):
...
+        """
+        Transfer the reference-pixel corrected data arrays from the
+        dataset to the input model
+
+        Parameters:
+        -----------
+
+        input_model: JWST datamodel
+
+        The datamodel the data arrays are to be transferred to
```
Comment (hbushouse) [inline jwst/refpix/reference_pixels.py]: "The datamodel to which the data arrays are to be transferred"  (I find it ironic that an Amurican is correcting a Brit's grammar ;-) )

File: jwst/refpix/reference_pixels.py [hunk 2]
```diff
@@ -181,6 +184,56 @@ def sigma_clip(self, data, dq, low=3.0, high=3.0):
...
+        embedded_subarray: nddata array
+        
+        The subarray embedded in the full-frame array
+        
+        """
+        rowstart = self.ystart - 1
+        rowstop = rowstart + self.ysize
+        colstart = self.xstart - 1
+        colstop = colstart + self.xsize
+        embedded_subarray = self.data[:, :, rowstart:rowstop, colstart:colstop]
```
Comment (hbushouse) [inline jwst/refpix/reference_pixels.py]: Is it ever possible to have the calculation of these slice indexes end up being outside the bounds of the actual array and hence we should implement some kind of checking to make sure that doesn't happen?

File: jwst/refpix/reference_pixels.py [hunk 3]
```diff
@@ -695,6 +755,61 @@ def do_corrections(self):
...
+        refdq = dqflags.pixel['REFERENCE_PIXEL']
+        (nints, ngroups, nrows, ncols) = self.data.shape
+        for integration in range(nints):
+            for group in range(ngroups):
+                #
+                # Get the reference values from the top and bottom reference
+                # pixels
+                #
+                thisgroup = self.data[integration, group].copy()
+                refpixindices = np.where(np.bitwise_and(self.pixeldq, refdq) == refdq)
```
Comment (hbushouse) [inline jwst/refpix/reference_pixels.py]: Where is it in the upstream processing that we're assuming the ``REFERENCE_PIXEL`` DQ flags got added to the ``pixeldq`` array? Are they supposed to come from the MASK reference file that gets loaded by the ``dq_init`` step? I don't recall if all MASK ref files have yet been updated to include those flags. Are the checks below sufficient to keep things from blowing up if those flags haven't been set?

Comment (stscirij) [reply jwst/refpix/reference_pixels.py]: Yes, I'm assuming they get added to the pixeldq array from the dq_init step.  If the flags haven't been set, the step should be skipped.

File: jwst/refpix/reference_pixels.py (hunk 2 above)
Comment (stscirij) [reply jwst/refpix/reference_pixels.py]: It's possible, but that would mean that data processing would have had to have introduced the error.  I'll put the check in, it's only a few extra lines.

File: jwst/refpix/reference_pixels.py [hunk 4]
```diff
@@ -194,7 +216,7 @@ def transfer_to_model(self, input_model):
 
         input_model: JWST datamodel
 
-        The datamodel the data arrays are to be transferred to
+        The datamodel to which the data arrays are to be transferred
```
Comment (hbushouse) [inline jwst/refpix/reference_pixels.py]: 👍

File: jwst/refpix/reference_pixels.py [hunk 5]
```diff
@@ -1327,6 +1365,16 @@ def do_left_right_correction(self, group, refvalues):
...
 
     def do_corrections(self):
+        if self.is_subarray:
+            self.do_subarray_corrections()
+        else:
+            self.do_fullframe_corrections()
+
+    def do_subarray_corrections(self):
+        log.warning("Refpix correction skipped for MIRI subarray")
+        return
```
Comment (hbushouse) [inline jwst/refpix/reference_pixels.py]: Does this in turn trigger the setting of the step completion status to "SKIPPED"?

Comment (stscirij) [reply jwst/refpix/reference_pixels.py]: Actually, this case is handled higher up now, so this line should never be needed.  The completion status will indeed be set to skipped for MIRI subarrays.

Comment (hbushouse) [reply jwst/refpix/reference_pixels.py]: 👍

**Pull request merged**